### DISCUSSION TITLE
3D to 2D projection

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#336] [Ideas]
i have been thinking about how to best project a 3D scene onto 2 dimensions - i.e. what would be the most convenient way to parametrize such a projection with parameters that are useful in an audio context - like in a 3D geometric synthesis where you can "modulate" the camera positioning parameters - stuff like that. i think, such a projection has in general 8 degrees of freedom: 

-camera position (x,y,z) - 3 values
-aiming point (x,y,z) - 3 values
-camera rotation (around the "aiming axis") - 1 value
-zoom - 1 value

makes 8 in total. have i missed anything? of course, the camera position could also be specified in spherical coordinates (distance, azimuth, elevation) and the "aiming point" could be replaced by an "aiming direction" with respect to the camera - there seem to be various ways of parametrizing it. at the end, i want to arrive at a simple projection equation that takes a 3D point (x,y,z) to a 2D point (X,Y) via:
```
 |X| = |xx xy xz| * |x| + |dx|
 |Y|   |yx yy yz|   |y|   |dy|
                    |z|
```
which has 8 coefficients as it should be for a 1-to-1 mapping being even possible. i hope this works out the way i think. am i on the right track?  if so, the task would be to figure out the formulas for the coefficients in terms of the user parameters. how are such projections defined in other 3D softwares? my computer graphics textbook specifies the view in terms of a lot more parameters - so i think, there must be redundancies in this - unless i'm missing something important which is entirely possible

**Comment by RobinSchmidt** [comment]
i think, specifying camera position in spherical coordinates is the best way in an audio context because it lets you set up periodic circulations of the camera around the object which should sound a bit like an LFO - periodic modulation - but not quite since it never reverses direction, so to speak. i generally tend to like cyclic parameters (i.e. angles) in the context of turning geometry into audio and modulate them with a sawtooth from -180 to +180. it's like a through-zero modulation or something

**Comment by elanhickler** [comment]
we need 3d snowflake

**Comment by RobinSchmidt** [comment]
> we need 3d snowflake

yeah! that would be awesome! i would need to write a 3D turtle graphics interpreter (should be straightforward). more difficult could become to create interesting L-system definitions in 3D. for 2D, there's a lot of stuff to grab from the web and books - for 3D, not so much - or i didn't find it yet. but in general - yes - a 3D snowflake is indeed something brewing in the back of my head.

**Comment by RobinSchmidt** [comment]
how about going up one more into 4D?

https://www.youtube.com/watch?v=iGO12Z5Lw8s

haha! ...but first things first!

**Comment by RobinSchmidt** [comment]
the problem with those multiple dimensions in this context is that you would need to define suitable turtle-graphics commands for rotations into the various directions. in 2D, there are only two possible directions (left/right), so + and - suffice. in 3D, you would (if i'm not mistaken) already need 6 symbols (left/right, up/down, forward/backward)...or however you want to call the directions (you need to rotate around x,y,z - where the z-rotations correspond to the 2D case, i.e. rotations that stay in the xy-plane if the input vector is in this plane) - but what characters should be used? at some point, you will run out of (meaningful) characters within the ascii set.

**Comment by RobinSchmidt** [comment]
...and in 4D - you don't rotate around an axis but instead around a plane (i still have to get my head around what this actually means) - so you would have many more combinations: xy, xz, xw, yz, yw, zw (*) - where xy, xz, yz correspond to the already known 3D rotations around z,y,x axes (always around the "missing variable" axis). ...a lot to digest - but interesting

(*) i think, the general rule is that for N dimensions, you would have binomial(N, N-2) possible independent rotations (how many independent axes are there in 3D = binomial(3,1) = 3, how many independent planes are there in 4D = binomial(4,2) = 6, etc.)
https://en.wikipedia.org/wiki/Combination
http://www.wolframalpha.com/input/?i=binomial(N,N-2)
= N * (N-1) / 2 so the number of independent rotations scales with the square of the number of dimensions - so it gets unmanagably big rather quickly -  but 3D and 4D might still be practical

**Comment by RobinSchmidt** [comment]
collecting some links:

https://github.com/abiusx/L3D
https://morphocode.com/3d-branching-structures-with-rabbit/
http://www.sccg.sk/~smolenova/elearning/ks_fmfiuk06.pdf
https://allenpike.com/modeling-plants-with-l-systems/
https://math.stackexchange.com/questions/123642/representing-a-3d-hilbert-curve-as-an-l-system

**Comment by RobinSchmidt** [comment]
back to 3D to 2D projections - i think, it would be best to adopt the same conventions as in OpenGL when setting these things up. here's something about how to set up a perspective projection:

http://www.scratchapixel.com/lessons/3d-basic-rendering/perspective-and-orthographic-projection-matrix/opengl-perspective-projection-matrix

and here's a lot more related and preliminary material:

http://www.scratchapixel.com/index.php

i probably don't need to go through all of it. my OpenGL book has stuff about that, too. all these things there are expressed in 4D homogeneous coordinates - ok, i guess, i can - as a final step - translate these representations into my target matrix-and-shift operation. actually, my matrix eqaution above is just a part of a 3D affine transform which would read in full:
```
 |X|   |xx xy xz|   |x|   |dx|
 |Y| = |yx yy yz| * |y| + |dy|
 |Z|   |zx zy zz|   |z|   |dz|
```
and then throw away the z coordinate - and as an optimization, i would not compute it in the first place, which is what my matrix equation in the original post shows (i think, i will set up my code such that it can be computed optionally for distance effects in other contexts). such 3D affine (= linear and shift) transforms can be represented as 4D linear transforms (i.e matrix multiplies) and that's why graphics APIs like this representation. translation of a transform from 4D-linear-in-homogeneous-coordinates to 3D-affine should be straightforward - i just need to look it up and make sure to get it right.

**Comment by RobinSchmidt** [comment]
hmmm...but this matrix seems to involve variables that i don't want to use in my parametrization...hmmm....we'll see

**Comment by RobinSchmidt** [comment]
i think, i made a mistake in my assumptions. after further reading, it appears to me that the equations for a perspective projection cannot be boiled down to a simple 3D affine transformation. you have this final "perspective division" step, which - as far as i can see - makes the whole transformation nonlinear. and it makes sense because a linear or affine transformation would preserve parallel lines which a perspective projection clearly doesn't (originally parallel lines in 3D converge to "vanishing points" in a perspective projection). so i guess, i'll have to use the 4D homogeneous coordinates approach for perspective projections. but perhaps my initial plan may still work out for parallel (aka orthographic) projections. i'll have to sit down and work out the formulas....later....for now, other stuff has higher priority

**Comment by RobinSchmidt** [comment]
btw. - in my OpenGL book (that i bought before you hired lorcan for the OpenGL stuff for prettyscope) i just stumbled over a formula that would let us specify a rotation not in terms of 3 basic rotations around the x,y,z axes (as is done in class rsRotationXYZ) but by specifying an rotation-axis vector and an angle. it's all about parametrization - you can do all kinds of rotations with both kinds of specifications. but the question is which one is more intuitive/useful/nicely-modulatable in a given context. making such decisions is the "art" part in such art-meets-science endeavours that we are dealing with. maybe i should implement a rotation class using this alternative specification

**Comment by RobinSchmidt** [comment]
i have a new book! this one:
https://www.amazon.de/gp/product/0387986820/
the presentation of projections is much more straightforward than in my other computer graphics book (this one: https://www.amazon.de/Computer-Graphics-Principles-James-Foley/dp/0201121107) and the dust is settling! i made some mistakes in my assumptions but i now have a pretty concrete plan in my mind how to work out the code for a general perspective (and parallel) projection. ...the book has even a little section on 4D to 3D projections. ...and nonlinear projections (fish eye, cylindric panorama, spherical (mercator, etc), stereographic, ..). and a short discussion of 3D turtle graphics. interesting and cool stuff, much of which may become relevant at some point!  will definitely be useful when going into 3D territory!

**Comment by RobinSchmidt** [comment]
i wonder if and how it could be possible to draw "solid" objects with sound instead of those wireframes that we currently have. like, in the simplest case, a rectangle of more or less uniform color (or gray shade). i think, we would have to let the virtual electron beam sort of "scan" the rectangle in a way similar to a CRT electron beam. just that we can't regulate the strength of the beam, so we would have to use the scanning speed as brightness control

**Comment by elanhickler** [comment]
Not sure what your intention is and how that would relate to sound. There's so much more beauty to explore with the simple beam concept, just waiting for PrettyScope development.

**Comment by RobinSchmidt** [comment]
> Not sure what your intention is and how that would relate to sound

that's what i'm tying to figure out myself. i've just been reading a couple of hundreds of pages of computer graphics and geometric modeling material (i'm now about two thirds through my new book). it's all cool math and algorithms stuff and i want to somehow apply it :-) but obviously, the book is written for people who want to implement the algorithmic backend for a 3D rendering software, so it's not directly applicable to audio. so i'm just brainstorming for ideas to apply the stuff from there in an audio context. i mean, wireframe models are obvious to translate to sound (although also not trivial - things like order of traversal of vertices, backtracking, etc. have to be dealt with).  i wonder if it's meaningfully possible to go beyond wireframe rendering.

btw. it also has 200+ pages about spline curves. bezier, b-spline, catmull-rom - a great reference for all the formulas, should we need them (could perhaps be relevant for the "oversampling" scope)

**Comment by RobinSchmidt** [comment]
i just had an idea, how to project any N-dimensional vector onto 2D:
![image](https://user-images.githubusercontent.com/1833099/42290872-5f332138-7fc8-11e8-875c-ef1173bed6ff.png)
it seems to be a pretty obvious thing to do -  however, i have never seen it being done this way before. it's very simple and it treats all coordinate axes in N-D on equal footing. note also, how the 3D projection reduces to the familiar y-axis-at-45°, z-axis-up projection when we choose k=3/4 (this would squish both 60° angles between x,y and y,z in the middle picture to 45°)

**Comment by RobinSchmidt** [comment]
> i have never seen it being done this way before

or - wait - actually i did - but only in the special case of 4D, in which case a tesseract (4D hypercube) would look like this:

![image](https://user-images.githubusercontent.com/1833099/42326870-a9257cbc-806a-11e8-8a74-9466f219c8f9.png)

**Comment by elanhickler** [comment]
you seen this?: https://www.youtube.com/watch?v=0t4aKJuKP0Q

**Comment by RobinSchmidt** [comment]
looks like an interesting app - but only available for ios and steam. this video seems to be a bit more revealing:

https://www.youtube.com/watch?v=S-yRYmdsnGs
looks like he's talking about this app? http://wenbolan.com/4dobject.html

anyway - it may be interesting to dive into the underlying math. maybe, it could be useful for oscilloscope music. do you know the book flatland? 

https://www.amazon.com/Flatland-Romance-Dimensions-Thrift-Editions/dp/048627263X

it's a classic, when it comes to 4D stuff. but even better is this one:

https://www.amazon.com/Geometry-Relativity-Fourth-Dimension-Mathematics/dp/0486234002

..a popular science book about 4D geometry and relativity - accessible to non-math people, but not totally stripped off all the formulas (i really find it annoying when popular science books leave out the nuts and bolts because the authors assume, the reader is unable to deal with stuff like an integral or a square-root - this author doesn't, which is refreshing). this book is, in fact, the one where i first saw this sort of rendering of the tesseract above (it's even on the cover).

**Comment by RobinSchmidt** [comment]
look at this:

https://www.youtube.com/watch?v=4URVJ3D8e8k&t=17m20s

from 17:20 to 18:00. i think, this would make for a pretty cool oscilloscope music synthesis algorithm. and the best thing about the 4D cube is: it should possible to traverse all the vertices visiting each one exactly once and then come back to the start vertex, i.e. no backtracking or jumping would be necessary in 4D - we could make a nice periodic signal from that. to synthesize a 3D cube - you would have to backtrack. i recently discovered the relevant math:

https://en.wikipedia.org/wiki/Eulerian_path

in the tesseract, each vertex has a degree of 4 (4 edges coming out of each vertex) - that's an even number, which implies, euler cycles exist. in a 3D cube, there can be no euler cycles. maybe, such an oscilloscope synth should provide more shapes - like whatever the 4D analogon of a tetraeder is (hypertetraeder? idk). the user could select the shape from a dropdown menu.
